# Information Retrieval 1#
## Assignment 2: Retrieval models [100 points] ##

In this assignment you will get familiar with basic and advanced information retrieval concepts. You will implement different information retrieval ranking models and evaluate their performance.

We provide you with a Indri index. To query the index, you'll use a Python package ([pyndri](https://github.com/cvangysel/pyndri)) that allows easy access to the underlying document statistics.

For evaluation you'll use the [TREC Eval](https://github.com/usnistgov/trec_eval) utility, provided by the National Institute of Standards and Technology of the United States. TREC Eval is the de facto standard way to compute Information Retrieval measures and is frequently referenced in scientific papers.

This is a **groups-of-three assignment**, the deadline is **Wednesday, January 31st**. Code quality, informative comments and convincing analysis of the results will be considered when grading. Submission should be done through blackboard, questions can be asked on the course [Piazza](piazza.com/university_of_amsterdam/spring2018/52041inr6y/home).

### Technicalities (must-read!) ###

The assignment directory is organized as follows:
   * `./assignment.ipynb` (this file): the description of the assignment.
   * `./index/`: the index we prepared for you.
   * `./ap_88_90/`: directory with ground-truth and evaluation sets:
      * `qrel_test`: test query relevance collection (**test set**).
      * `qrel_validation`: validation query relevance collection (**validation set**).
      * `topics_title`: semicolon-separated file with query identifiers and terms.

You will need the following software packages (tested with Python 3.5 inside [Anaconda](https://conda.io/docs/user-guide/install/index.html)):
   * Python 3.5 and Jupyter
   * Indri + Pyndri (Follow the installation instructions [here](https://github.com/nickvosk/pyndri/blob/master/README.md))
   * gensim [link](https://radimrehurek.com/gensim/install.html)
   * TREC Eval [link](https://github.com/usnistgov/trec_eval)

### TREC Eval primer ###
The TREC Eval utility can be downloaded and compiled as follows:

    git clone https://github.com/usnistgov/trec_eval.git
    cd trec_eval
    make

TREC Eval computes evaluation scores given two files: ground-truth information regarding relevant documents, named *query relevance* or *qrel*, and a ranking of documents for a set of queries, referred to as a *run*. The *qrel* will be supplied by us and should not be changed. For every retrieval model (or combinations thereof) you will generate a run of the top-1000 documents for every query. The format of the *run* file is as follows:

    $query_identifier Q0 $document_identifier $rank_of_document_for_query $query_document_similarity $run_identifier
    
where
   * `$query_identifier` is the unique identifier corresponding to a query (usually this follows a sequential numbering).
   * `Q0` is a legacy field that you can ignore.
   * `$document_identifier` corresponds to the unique identifier of a document (e.g., APXXXXXXX where AP denotes the collection and the Xs correspond to a unique numerical identifier).
   * `$rank_of_document_for_query` denotes the rank of the document for the particular query. This field is ignored by TREC Eval and is only maintained for legacy support. The ranks are computed by TREC Eval itself using the `$query_document_similarity` field (see next). However, it remains good practice to correctly compute this field.
   * `$query_document_similarity` is a score indicating the similarity between query and document where a higher score denotes greater similarity.
   * `$run_identifier` is an identifier of the run. This field is for your own convenience and has no purpose beyond bookkeeping.
   
For example, say we have two queries: `Q1` and `Q2` and we rank three documents (`DOC1`, `DOC2`, `DOC3`). For query `Q1`, we find the following similarity scores `score(Q1, DOC1) = 1.0`, `score(Q1, DOC2) = 0.5`, `score(Q1, DOC3) = 0.75`; and for `Q2`: `score(Q2, DOC1) = -0.1`, `score(Q2, DOC2) = 1.25`, `score(Q1, DOC3) = 0.0`. We can generate run using the following snippet:

In [1]:
import logging
import sys
import os
import time
import math
import numpy as np
import subprocess
import functools
import random
from scipy import stats


def write_run(model_name, data, out_f,
              max_objects_per_query=sys.maxsize,
              skip_sorting=False):
    """
    Write a run to an output file.
    Parameters:
        - model_name: identifier of run.
        - data: dictionary mapping topic_id to object_assesments;
            object_assesments is an iterable (list or tuple) of
            (relevance, object_id) pairs.
            The object_assesments iterable is sorted by decreasing order.
        - out_f: output file stream.
        - max_objects_per_query: cut-off for number of objects per query.
    """
    for subject_id, object_assesments in data.items():
        if not object_assesments:
            logging.warning('Received empty ranking for %s; ignoring.',
                            subject_id)

            continue

        # Probe types, to make sure everything goes alright.
        # assert isinstance(object_assesments[0][0], float) or \
        #     isinstance(object_assesments[0][0], np.float32)
        assert isinstance(object_assesments[0][1], str) or \
            isinstance(object_assesments[0][1], bytes)

        if not skip_sorting:
            object_assesments = sorted(object_assesments, reverse=True)

        if max_objects_per_query < sys.maxsize:
            object_assesments = object_assesments[:max_objects_per_query]

        if isinstance(subject_id, bytes):
            subject_id = subject_id.decode('utf8')

        for rank, (relevance, object_id) in enumerate(object_assesments):
            if isinstance(object_id, bytes):
                object_id = object_id.decode('utf8')

            out_f.write(
                '{subject} Q0 {object} {rank} {relevance} '
                '{model_name}\n'.format(
                    subject=subject_id,
                    object=object_id,
                    rank=rank + 1,
                    relevance=relevance,
                    model_name=model_name))
            
# The following writes the run to standard output.
# In your code, you should write the runs to local
# storage in order to pass them to trec_eval.
write_run(
    model_name='example',
    data={
        'Q1': ((1.0, 'DOC1'), (0.5, 'DOC2'), (0.75, 'DOC3')),
        'Q2': ((-0.1, 'DOC1'), (1.25, 'DOC2'), (0.0, 'DOC3')),
    },
    out_f=sys.stdout,
    max_objects_per_query=1000)

Q2 Q0 DOC2 1 1.25 example
Q2 Q0 DOC3 2 0.0 example
Q2 Q0 DOC1 3 -0.1 example
Q1 Q0 DOC1 1 1.0 example
Q1 Q0 DOC3 2 0.75 example
Q1 Q0 DOC2 3 0.5 example


Now, imagine that we know that `DOC1` is relevant and `DOC3` is non-relevant for `Q1`. In addition, for `Q2` we only know of the relevance of `DOC3`. The query relevance file looks like:

    Q1 0 DOC1 1
    Q1 0 DOC3 0
    Q2 0 DOC3 1
    
We store the run and qrel in files `example.run` and `example.qrel` respectively on disk. We can now use TREC Eval to compute evaluation measures. In this example, we're only interested in Mean Average Precision and we'll only show this below for brevity. However, TREC Eval outputs much more information such as NDCG, recall, precision, etc.

    $ trec_eval -m all_trec -q example.qrel example.run | grep -E "^map\s"
    > map                   	Q1	1.0000
    > map                   	Q2	0.5000
    > map                   	all	0.7500
    
Now that we've discussed the output format of rankings and how you can compute evaluation measures from these rankings, we'll now proceed with an overview of the indexing framework you'll use.

### Pyndri primer ###
For this assignment you will use [Pyndri](https://github.com/cvangysel/pyndri) [[1](https://arxiv.org/abs/1701.00749)], a python interface for [Indri](https://www.lemurproject.org/indri.php). We have indexed the document collection and you can query the index using Pyndri. We will start by giving you some examples of what Pyndri can do:

First we read the document collection index with Pyndri:

In [2]:
import pyndri

index = pyndri.Index('index/')

The loaded index can be used to access a collection of documents in an easy manner. We'll give you some examples to get some idea of what it can do, it is up to you to figure out how to use it for the remainder of the assignment.

First let's look at the number of documents, since Pyndri indexes the documents using incremental identifiers we can simply take the lowest index and the maximum document and consider the difference:

In [3]:
print("There are %d documents in this collection." % (index.maximum_document() - index.document_base()))

There are 164597 documents in this collection.


Let's take the first document out of the collection and take a look at it:

In [4]:
example_document = index.document(index.document_base())
print(example_document)

('AP890425-0001', (1360, 192, 363, 0, 880, 0, 200, 0, 894, 412, 92160, 3, 192, 0, 363, 34, 1441, 0, 174134, 0, 200, 0, 894, 412, 2652, 0, 810, 107, 49, 4903, 420, 0, 1, 48, 35, 489, 0, 35, 687, 192, 243, 0, 249311, 1877, 0, 1651, 1174, 0, 2701, 117, 412, 0, 810, 391, 245233, 1225, 5838, 16, 0, 233156, 3496, 0, 393, 17, 0, 2435, 4819, 930, 0, 0, 200, 0, 894, 0, 22, 398, 145, 0, 3, 271, 115, 0, 1176, 2777, 292, 0, 725, 192, 0, 0, 50046, 0, 1901, 1130, 0, 192, 0, 408, 0, 243779, 0, 0, 553, 192, 0, 363, 0, 3747, 0, 0, 0, 0, 1176, 0, 1239, 0, 0, 1115, 17, 0, 0, 585, 192, 1963, 0, 0, 412, 54356, 0, 773, 0, 0, 0, 192, 0, 0, 1130, 0, 363, 0, 545, 192, 0, 1174, 1901, 1130, 0, 4, 398, 145, 39, 0, 577, 0, 355, 0, 491, 0, 6025, 0, 0, 193156, 88, 34, 437, 0, 0, 1852, 0, 828, 0, 1588, 0, 0, 0, 2615, 0, 0, 107, 49, 420, 0, 0, 190, 7, 714, 2701, 0, 237, 192, 157, 0, 412, 34, 437, 0, 0, 200, 6025, 26, 0, 0, 0, 0, 363, 0, 22, 398, 145, 0, 200, 638, 126222, 6018, 0, 880, 0, 0, 161, 0, 0, 319, 894, 2701, 

Here we see a document consists of two things, a string representing the external document identifier and an integer list representing the identifiers of words that make up the document. Pyndri uses integer representations for words or terms, thus a token_id is an integer that represents a word whereas the token is the actual text of the word/term. Every id has a unique token and vice versa with the exception of stop words: words so common that there are uninformative, all of these receive the zero id.

To see what some ids and their matching tokens we take a look at the dictionary of the index:

In [5]:
token2id, id2token, _ = index.get_dictionary()
print(list(id2token.items())[:15])

[(1, 'new'), (2, 'percent'), (3, 'two'), (4, '1'), (5, 'people'), (6, 'million'), (7, '000'), (8, 'government'), (9, 'president'), (10, 'years'), (11, 'state'), (12, '2'), (13, 'states'), (14, 'three'), (15, 'time')]


Using this dictionary we can see the tokens for the (non-stop) words in our example document:

In [6]:
print([id2token[word_id] for word_id in example_document[1] if word_id > 0])

['52', 'students', 'arrested', 'takeover', 'university', 'massachusetts', 'building', 'fifty', 'two', 'students', 'arrested', 'tuesday', 'evening', 'occupying', 'university', 'massachusetts', 'building', 'overnight', 'protest', 'defense', 'department', 'funded', 'research', 'new', 'york', 'city', 'thousands', 'city', 'college', 'students', 'got', 'unscheduled', 'holiday', 'demonstrators', 'occupied', 'campus', 'administration', 'building', 'protest', 'possible', 'tuition', 'increases', 'prompting', 'officials', 'suspend', 'classes', '60', 'police', 'riot', 'gear', 'arrived', 'university', 'massachusetts', '5', 'p', 'm', 'two', 'hours', 'later', 'bus', 'drove', 'away', '29', 'students', 'camped', 'memorial', 'hall', 'students', 'charged', 'trespassing', '23', 'students', 'arrested', 'lying', 'bus', 'prevent', 'leaving', 'police', '300', 'students', 'stood', 'building', 'chanting', 'looking', 'students', 'hall', 'arrested', '35', 'students', 'occupied', 'memorial', 'hall', '1', 'p', 'm',

The reverse can also be done, say we want to look for news about the "University of Massachusetts", the tokens of that query can be converted to ids using the reverse dictionary:

In [7]:
query_tokens = index.tokenize("University of Massachusetts")
print("Query by tokens:", query_tokens)
query_id_tokens = [token2id.get(query_token,0) for query_token in query_tokens]
print("Query by ids with stopwords:", query_id_tokens)
query_id_tokens = [word_id for word_id in query_id_tokens if word_id > 0]
print("Query by ids without stopwords:", query_id_tokens)

Query by tokens: ['university', '', 'massachusetts']
Query by ids with stopwords: [200, 0, 894]
Query by ids without stopwords: [200, 894]


Naturally we can now match the document and query in the id space, let's see how often a word from the query occurs in our example document:

In [8]:
matching_words = sum([True for word_id in example_document[1] if word_id in query_id_tokens])
print("Document %s has %d word matches with query: \"%s\"." % (example_document[0], matching_words, ' '.join(query_tokens)))
print("Document %s and query \"%s\" have a %.01f%% overlap." % (example_document[0], ' '.join(query_tokens),matching_words/float(len(example_document[1]))*100))

Document AP890425-0001 has 13 word matches with query: "university  massachusetts".
Document AP890425-0001 and query "university  massachusetts" have a 2.5% overlap.


While this is certainly not everything Pyndri can do, it should give you an idea of how to use it. Please take a look at the [examples](https://github.com/cvangysel/pyndri) as it will help you a lot with this assignment.

**CAUTION**: Avoid printing out the whole index in this Notebook as it will generate a lot of output and is likely to corrupt the Notebook.

### Parsing the query file
You can parse the query file (`ap_88_89/topics_title`) using the following snippet:

In [9]:
import collections
import io
import logging
import sys

def parse_topics(file_or_files,
                 max_topics=sys.maxsize, delimiter=';'):
    assert max_topics >= 0 or max_topics is None

    topics = collections.OrderedDict()

    if not isinstance(file_or_files, list) and \
            not isinstance(file_or_files, tuple):
        if hasattr(file_or_files, '__iter__'):
            file_or_files = list(file_or_files)
        else:
            file_or_files = [file_or_files]

    for f in file_or_files:
        assert isinstance(f, io.IOBase)

        for line in f:
            assert(isinstance(line, str))

            line = line.strip()

            if not line:
                continue

            topic_id, terms = line.split(delimiter, 1)

            if topic_id in topics and (topics[topic_id] != terms):
                    logging.error('Duplicate topic "%s" (%s vs. %s).',
                                  topic_id,
                                  topics[topic_id],
                                  terms)

            topics[topic_id] = terms

            if max_topics > 0 and len(topics) >= max_topics:
                break

    return topics

with open('./ap_88_89/topics_title', 'r') as f_topics:
    print(parse_topics([f_topics]))

OrderedDict([('51', 'Airbus Subsidies'), ('52', 'South African Sanctions'), ('53', 'Leveraged Buyouts'), ('54', 'Satellite Launch Contracts'), ('55', 'Insider Trading'), ('56', 'Prime (Lending) Rate Moves, Predictions'), ('57', 'MCI'), ('58', 'Rail Strikes'), ('59', 'Weather Related Fatalities'), ('60', 'Merit-Pay vs. Seniority'), ('61', 'Israeli Role in Iran-Contra Affair'), ('62', "Military Coups D'etat"), ('63', 'Machine Translation'), ('64', 'Hostage-Taking'), ('65', 'Information Retrieval Systems'), ('66', 'Natural Language Processing'), ('67', 'Politically Motivated Civil Disturbances'), ('68', 'Health Hazards from Fine-Diameter Fibers'), ('69', 'Attempts to Revive the SALT II Treaty'), ('70', 'Surrogate Motherhood'), ('71', 'Border Incursions'), ('72', 'Demographic Shifts in the U.S.'), ('73', 'Demographic Shifts across National Boundaries'), ('74', 'Conflicting Policy'), ('75', 'Automation'), ('76', 'U.S. Constitution - Original Intent'), ('77', 'Poaching'), ('78', 'Greenpeace'

### Task 1: Implement and compare lexical IR methods [35 points] ### 

In this task you will implement a number of lexical methods for IR using the **Pyndri** framework. Then you will evaluate these methods on the dataset we have provided using **TREC Eval**.

Use the **Pyndri** framework to get statistics of the documents (term frequency, document frequency, collection frequency; **you are not allowed to use the query functionality of Pyndri**) and implement the following scoring methods in **Python**:

- [TF-IDF](http://nlp.stanford.edu/IR-book/html/htmledition/tf-idf-weighting-1.html) and 
- [BM25](http://nlp.stanford.edu/IR-book/html/htmledition/okapi-bm25-a-non-binary-model-1.html) with k1=1.2 and b=0.75. **[5 points]**
- Language models ([survey](https://drive.google.com/file/d/0B-zklbckv9CHc0c3b245UW90NE0/view))
    - Jelinek-Mercer (explore different values of 𝛌 in the range [0.1, 0.5, 0.9]). **[5 points]**
    - Dirichlet Prior (explore different values of 𝛍 [500, 1000, 1500]). **[5 points]**
    - Absolute discounting (explore different values of 𝛅 in the range [0.1, 0.5, 0.9]). **[5 points]**
    - [Positional Language Models](http://sifaka.cs.uiuc.edu/~ylv2/pub/sigir09-plm.pdf) define a language model for each position of a document, and score a document based on the scores of its PLMs. The PLM is estimated based on propagated counts of words within a document through a proximity-based density function, which both captures proximity heuristics and achieves an effect of “soft” passage retrieval. Implement the PLM, all five kernels, but only the Best position strategy to score documents. Use 𝛔 equal to 50, and Dirichlet smoothing with 𝛍 optimized on the validation set (decide how to optimize this value yourself and motivate your decision in the report). **[10 points]**
    
Implement the above methods and report evaluation measures (on the test set) using the hyper parameter values you optimized on the validation set (also report the values of the hyper parameters). Use TREC Eval to obtain the results and report on `NDCG@10`, Mean Average Precision (`MAP@1000`), `Precision@5` and `Recall@1000`.

For the language models, create plots showing `NDCG@10` with varying values of the parameters. You can do this by chaining small scripts using shell scripting (preferred) or execute trec_eval using Python's `subprocess`.

Compute significance of the results using a [two-tailed paired Student t-test](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_rel.html) **[5 points]**. Be wary of false rejection of the null hypothesis caused by the [multiple comparisons problem](https://en.wikipedia.org/wiki/Multiple_comparisons_problem). There are multiple ways to mitigate this problem and it is up to you to choose one.

Analyse the results by identifying specific queries where different methods succeed or fail and discuss possible reasons that cause these differences. This is *very important* in order to understand who the different retrieval functions behave.

**NOTE**: Don’t forget to use log computations in your calculations to avoid underflows. 

**IMPORTANT**: You should structure your code around the helper functions we provide below.

In [10]:
with open('./ap_88_89/topics_title', 'r') as f_topics:
    queries = parse_topics([f_topics])

index = pyndri.Index('index/')
# the total number of documents
num_documents = index.maximum_document() - index.document_base()

# the pyindri dictionary
dictionary = pyndri.extract_dictionary(index)

# tokenize the queries
# query_id --> [token_id1, token_id2, ...] 
tokenized_queries = {
    query_id: [dictionary.translate_token(token)
               for token in index.tokenize(query_string)
               if dictionary.has_token(token)]
    for query_id, query_string in queries.items()}

# gather all of the query token_ids into a set TODO why? 
query_term_ids = set(
    query_term_id
    for query_term_ids in tokenized_queries.values()
    for query_term_id in query_term_ids)

print('Gathering statistics about', len(query_term_ids), 'terms.')

#### Inverted index creation. 
# per document dictionaries
document_lengths = {} # doc_id -> length of document
unique_terms_per_document = {} # doc_id -> number of unique tokens in document

# (query_term, (doc_id)) -> document TF
inverted_index = collections.defaultdict(dict)
# (query_term_id) -> frequency of occurence of query_term_id in all of the documents 
collection_frequencies = collections.defaultdict(int)
# total number of words in the document
total_terms = 0

start_time = time.time()
doc_id_to_ext_id = {}

for int_doc_id in range(index.document_base(), index.maximum_document()):
    ext_doc_id, doc_token_ids = index.document(int_doc_id)
    doc_id_to_ext_id[int_doc_id] = ext_doc_id
     
    document_bow = collections.Counter(
        token_id for token_id in doc_token_ids
        if token_id > 0)
    
    
    document_length = sum(document_bow.values())

    document_lengths[int_doc_id] = document_length
    total_terms += document_length

    unique_terms_per_document[int_doc_id] = len(document_bow)
    
    # accumulate TF and CF
    for query_term_id in query_term_ids:
        assert query_term_id is not None

        document_term_frequency = document_bow.get(query_term_id, 0)

        if document_term_frequency == 0:
            continue

        collection_frequencies[query_term_id] += document_term_frequency
        inverted_index[query_term_id][int_doc_id] = document_term_frequency
    
avg_doc_length = total_terms / num_documents

print('Inverted index creation took', time.time() - start_time, 'seconds.')

Gathering statistics about 456 terms.
Inverted index creation took 29.279672384262085 seconds.


In [12]:
def run_retrieval(model_name, score_fn):
    """
    Runs a retrieval method for all the queries and writes the TREC-friendly results in a file.
    
    :param model_name: the name of the model (a string)
    :param score_fn: the scoring function (a function - see below for an example) 
    """
    run_out_path = '{}.run'.format(model_name)

    #if os.path.exists(run_out_path):
    #    return

    retrieval_start_time = time.time()

    print('Retrieving using', model_name)
    
    
    score_sums = collections.defaultdict(lambda: collections.defaultdict(float))
    
    for query_id, query_terms in queries.items():
        for query_term_id in tokenized_queries[query_id]:
            # TODO what happens when the query term doesn't occur in any document
            for doc_id, document_term_freq in inverted_index[query_term_id].items():
                score = score_fn(doc_id, query_term_id, document_term_freq)
                score_sums[query_id][doc_id] += score
                
    data = {}
    for query_id, document_scores in score_sums.items():
        doc_scores = []
        for doc_id in document_scores.keys():
            doc_scores.append((document_scores[doc_id], doc_id_to_ext_id[doc_id]))
        doc_scores.sort(key = lambda _: -_[0])
        data[query_id] = doc_scores
    
    with open(run_out_path, 'w') as f_out:
        write_run(
            model_name=model_name,
            data=data,
            out_f=f_out,
            max_objects_per_query=1000)

In [13]:
def tfidf(int_document_id, query_term_id, document_term_freq):
    """
    Scoring function for a document and a query term
    
    :param int_document_id: the document id
    :param query_token_id: the query term id (assuming you have split the query to tokens)
    :param document_term_freq: the document term frequency of the query term 
    """
    idf = math.log(num_documents / len(inverted_index[query_term_id]))
    tf = math.log(1 + document_term_freq)
    return tf * idf

In [14]:
def bm25(int_document_id, query_term_id, document_term_freq, k_1 = 1.2, b = 0.75):
    """
    TODO
    """
    idf = math.log(num_documents / len(inverted_index[query_term_id]))
    tf_numerator = (k_1 + 1) * document_term_freq 
    tf_denom = k_1 * ((1 - b) + b * (document_lengths[int_document_id] / avg_doc_length)) + document_term_freq
    return idf * (tf_numerator / tf_denom)

In [15]:
def jelinek_mercer(int_document_id, query_term_id, document_term_freq, lmbd = 0.5):
    """
    TODO
    """
    document_contrib = lmbd * (document_term_freq / document_lengths[int_document_id]) 
    corpus_contrib = (1 - lmbd) * (collection_frequencies[query_term_id] / total_terms)
    return np.log(document_contrib + corpus_contrib)

In [16]:
def dirichlet_prior(int_document_id, query_term_id, document_term_freq, mu = 1000):
    """
    TODO
    """
    #print("mu", mu)
    numerator = document_term_freq + mu * (collection_frequencies[query_term_id] / total_terms)
    denom = document_lengths[int_document_id] + mu
    return np.log(numerator / denom)

In [17]:
def absolute_discounting(int_document_id, query_term_id, document_term_freq, delta = 0.1):
    """
    sigma: discount constant [0,1]
    delta: sigma*|du|/|d|
    """
    sigma = (delta * unique_terms_per_document[int_document_id]) / document_lengths[int_document_id]
    numerator = max(document_term_freq - delta, 0) 
    denominator = document_lengths[int_document_id]
    discounted_prob = numerator / denominator
    unigram_prob = collection_frequencies[query_term_id] / total_terms
    return np.log(discounted_prob + (sigma * unigram_prob))

In [18]:
 #please check, by ece

def gaussian_kernel(i, j, sigma):
    return np.exp(-np.power((i-j),2)/(2*np.power(sigma,2)))

def gaussian_kernel_estimation(i, j, sigma, N):
    #TODO the one with the cumulative density function CDF
    pass
    
def triangle_kernel(i, j, sigma): 
    absolute = np.absolute(i-j)
    return 1 - (absolute/sigma) if(absolute <= sigma) else 0

def cosine_kernel(i, j, sigma):
    absolute = np.absolute(i-j) 
    k = 0.0
    
    if absolute <= sigma:
        k = (1 + np.cos((absolute*math.pi)/sigma))/2
    
    return k

def circle_kernel(i, j, sigma):
    absolute = np.absolute(i-j) 
    k = 0.0
    
    if absolute <= sigma:
        k = np.sqrt(1 - np.power(absolute/sigma, 2))
    
    return k

def passage_kernel(i, j, sigma):
    absolute = np.absolute(i-j)
    return 1 if(absolute <= sigma) else 0


In [52]:
def run_retrieval_plm(model_name, score_fn):
    
    #please check, by ece
    
    """
    Runs a retrieval method for all the queries and writes the TREC-friendly results in a file.
    
    :param model_name: the name of the model (a string)
    :param score_fn: the scoring function (a function - see below for an example) 
    """
    run_out_path = '{}.run'.format(model_name)

    #if os.path.exists(run_out_path):
    #    return

    retrieval_start_time = time.time()

    print('Retrieving using', model_name)
    
    
    score_sums = collections.defaultdict(lambda: collections.defaultdict(float))
    
    plm_models_wdi = collections.defaultdict(lambda: collections.defaultdict(lambda: collections.defaultdict(float)))
            
    scores_qdi = collections.defaultdict(lambda: collections.defaultdict(lambda: collections.defaultdict(float)))
                
    
    #MLE probs for query_lm[query_id][query_term_id]
    query_lm = collections.defaultdict(lambda: collections.defaultdict(float)) 
    
    for query_id, query_terms in queries.items():
        query_word_counts = collections.defaultdict(float)
        tokenized_qs = tokenized_queries[query_id]
        for query_term_id in tokenized_qs:
            query_word_counts[query_term_id] += 1 #P(w_i|q)
         
        query_word_counts.update((k,v/len(tokenized_qs)) for k,v in query_word_counts.items())#this was buggy, changed it
       
        query_lm[query_id] = query_word_counts #for each query, keep LM = P(w|q) for each unique word in the query
    
    doc_tokens_with_stop = {}
    
    # extract tokens without stop words
    for int_doc_id in range(index.document_base(), index.maximum_document()):
        doc_tokens_with_stop[int_doc_id] = [ind for ind in index.document(int_doc_id)[1]]
    
    for query_id, query_terms in queries.items():
        for query_term_id in tokenized_queries[query_id]:
            # TODO what happens when the query term doesn't occur in any document
            for doc_id, document_term_freq in inverted_index[query_term_id].items():
              
                #positional LM
                plm_models_wdi[query_term_id][doc_id] = score_fn(doc_id, query_term_id, document_term_freq,
                                                                 doc_tokens_with_stop[doc_id])
                q_lm = query_lm[query_id][query_term_id]
                
                full_doc_length = len(doc_tokens_with_stop[doc_id])
                
                for i in range(full_doc_length):
                    #for one word
                    
                    #final scores given query, doc and index
                    scores_qdi[query_id][doc_id][i] -= q_lm * np.log(q_lm / plm_models_wdi[query_term_id][doc_id][i]) 
                    #print(i,scores_qdi[query_id][doc_id][i])
               
                #scores_sum SHOULD BE RETURNED as the scores of each doc
                #print(max(list(scores_qdi[query_id][doc_id].values())))
                
                #oprint(score_sums[query_id][doc_id])
                score_sums[query_id][doc_id] = max(list(scores_qdi[query_id][doc_id].values()))
                #print(query_id, doc_id,doc_id_to_ext_id[doc_id], score_sums[query_id][doc_id] )
                
    data = {}
    for query_id, document_scores in score_sums.items():
        doc_scores = []
        for doc_id in document_scores.keys():
            doc_scores.append((document_scores[doc_id], doc_id_to_ext_id[doc_id]))
        doc_scores.sort(key = lambda _: -_[0])
        data[query_id] = doc_scores
    
    with open(run_out_path, 'w') as f_out:
        write_run(
            model_name=model_name,
            data=data,
            out_f=f_out,
            max_objects_per_query=1000)

In [55]:
def PLM(int_document_id, query_term_id, document_term_freq, indices, kernel = gaussian_kernel, sigma = 50, mu = 1500):

    #please check, by ece
    
    #change kernel?
    #change mu according to optimized dirichlet hyperparameter!
    #sigma is given in the assignment
    
    #query_term_id = w
    #c_prime is for w,i
    
    full_doc_length = len(indices)
    
    c_prime = np.zeros(full_doc_length)
    
    #get the positional word ids in the document, eliminate stop words where ind==0
    #TODO i and j after removing 0 or are we keeping the indices from the actual sequence?
    #WE ARE KEEPING EVERYTHING IN THE DOCUMENT
     
    
    #denominator, virtual document lengths
    z = np.zeros(full_doc_length)
    
    for i in range(full_doc_length):
        
        #kernel outcomes for the given i over all j, they will be summed up for z[i]
        k = np.zeros(full_doc_length)
        
        for j in range(full_doc_length):
            
            cwj = 0.0
            if indices[j] == query_term_id:
                cwj = 1.0
            
            k[j] = kernel(i,j, sigma) #TODO ij or ji
            c_prime[i] += cwj*k[j]
            
        z[i] = np.sum(k)   #sum over kernel(j,i) j=1 to N, not the final simplified version
        #NOT USING the sum of c_prime over all w, but using zi as the sum over kernels
        
    unigram_prob = collection_frequencies[query_term_id] / total_terms
    
    pm_wdi = (c_prime + mu * unigram_prob)/(z + mu)
    
    #language model per word given document and a position
    return pm_wdi

In [56]:
# combining the two functions above: 
#run_retrieval('tfidf', tfidf)
#run_retrieval('bm25', bm25)
#run_retrieval('jelinek_mercer', jelinek_mercer)
#run_retrieval('dirichlet_prior', dirichlet_prior)
#run_retrieval('absolute_discounting', absolute_discounting)
run_retrieval_plm('PLM', PLM)
# TODO implement the rest of the retrieval functions 

# TODO implement tools to help you with the analysis of the results.

Retrieving using PLM
51 39937 AP891002-0137 -3.32710682308
51 108551 AP880915-0047 -3.31976389652
51 141472 AP880706-0025 -3.06194641513
51 76813 AP891017-0091 -3.3326605372
51 96270 AP880805-0037 -3.07922249955
51 143375 AP881222-0148 -3.33263931017
51 107011 AP880820-0096 -3.06398468725
51 131092 AP880804-0167 -2.64648918325
51 114710 AP880906-0086 -3.3208541051
51 25623 AP890906-0311 -3.32554842467
51 54298 AP890810-0244 -3.33270464798
51 116772 AP880312-0117 -3.33265750279
51 100389 AP881020-0170 -3.32398635083
51 138278 AP881231-0042 -3.32983504572
51 141660 AP880706-0213 -3.31768121141
51 77866 AP890131-0301 -3.33270454178
51 95275 AP880311-0301 -2.97707556526
51 31789 AP890609-0270 -3.33174437555
51 93533 AP881230-0079 -3.33269870393
51 19516 AP890517-0089 -3.14001462084
51 78911 AP891219-0213 -3.33270464616
51 65602 AP890125-0094 -3.32249208381
51 20548 AP890216-0288 -2.78288290298
51 76470 AP890608-0253 -2.72581865045
51 81990 AP890110-0304 -3.00101634377
51 17480 AP890614-028

51 119295 AP880710-0042 -3.33171718467
51 16896 AP890911-0278 -3.32483573755
51 53763 AP891222-0290 -3.33270433405
51 107013 AP880820-0098 -3.31344627063
51 132616 AP880712-0223 -3.33134737178
51 46607 AP890411-0191 -3.32950000879
51 136511 AP880325-0293 -2.64654716315
51 82450 AP891103-0231 -3.33134774951
51 75284 AP890103-0117 -3.33270337528
51 71192 AP891117-0207 -3.33269776836
51 134685 AP880411-0118 -3.32653898375
51 71198 AP891117-0213 -2.77120536482
51 71199 AP891117-0214 -2.95334996638
51 123427 AP880709-0167 -3.31359158128
51 119333 AP880710-0080 -3.33269601197
51 81330 AP890427-0258 -3.33270464866
51 123438 AP880709-0178 -3.32950000879
51 39473 AP890413-0262 -3.33198215336
51 122419 AP880827-0035 -3.33269533902
51 130613 AP880627-0304 -3.32950000879
51 141661 AP880706-0214 -3.33196288347
51 96823 AP880908-0237 -3.33270464867
51 126521 AP881019-0150 -3.32871128642
51 72252 AP890104-0259 -2.73326489278
51 91712 AP880229-0271 -3.33261973214
51 112195 AP880628-0087 -3.31859016827

KeyboardInterrupt: 

In [18]:

class TrecAPI:
    def __init__(self, trec_path = "../../trec_eval/trec_eval"):
        self.trec_path = os.path.abspath(trec_path)
    
    def evaluate(self, test_file_name, prediction_file_name, metrics_to_capture=None, granular = False):
        """
        TODO
        """
        if metrics_to_capture is None:
            metrics_to_capture = {"ndcg_cut_10", "map_cut_1000", "P_5", "recall_1000"}
        try:
            command = [self.trec_path, "-m", "all_trec", "-q", test_file_name, prediction_file_name]
            output = subprocess.check_output(command, universal_newlines=True)
            data = collections.defaultdict(dict)
            for line in output.split("\n"):
                # ignore empty lines
                if line.strip() == "":
                    continue
                    
                metric, query, value = line.split("\t")
                
                if not granular and query != "all":
                    continue
                    
                
                metric = metric.strip()
                    
                # ignore metrics we don't care about
                if metric not in metrics_to_capture:
                    continue
                
                # relstring is a binary string, don't convert to float
                if metric not in {"relstring"}:
                    value = float(value)
                
                data[query][metric] = value 
            
            if not granular:
                return data["all"]
            
            return data
        except subprocess.CalledProcessError as e:
            print(e.output)
            return None
        

trec = TrecAPI()
results = trec.evaluate("./ap_88_89/qrel_validation", "./dirichlet_prior.run")
results

{'P_5': 0.2133,
 'map_cut_1000': 0.173,
 'ndcg_cut_10': 0.2382,
 'recall_1000': 0.5728}

In [19]:
def grid_eval(grid, score_fn, test_file, verbose = False, output_folder = "grid_results/"):
    """
    Evaluate a scoring method on a grid of possible values.
    grid : a dictionary of `param -> list of possible values`
    scoring_method: the scoring method to evaluate
    returns TODO
    """
    # convert to an list because indexing is required
    grid = list(grid.items())
    param_lengths = tuple([len(grid_item[1]) for grid_item in grid])
    print("Running grid eval on {} possible combinations".format(np.prod(param_lengths)))
    
    all_indices = np.ndindex(param_lengths)
    grid_results = []
    for index, values in enumerate(all_indices):
        param_grid = {}
        for param_index, param_val_index in enumerate(values):
            param_name, param_values = grid[param_index]
            param_grid[param_name] = param_values[param_val_index]
        if verbose:
            print("Running evaluation {} of {}, with params: {}".format(index + 1, np.prod(param_lengths), param_grid))
        output_path = os.path.join(output_folder, "{}_{}".format(score_fn.__name__, index))
        # construct the scoring function
        part_score_fn = functools.partial(score_fn, **param_grid) 
        run_retrieval(output_path, part_score_fn)
        metrics = trec.evaluate(test_file, output_path + ".run")
        grid_results.append((param_grid, metrics))
    return grid_results

##### Parameter tuning for JM
Jelinek-Mercer (explore different values of 𝛌 in the range [0.1, 0.5, 0.9]).

In [23]:
grid_jm = {
    "lmbd" : np.linspace(0.1, 0.9, 10)
}
jelinek_mercer_grid_results = grid_eval(grid_jm, jelinek_mercer, "./ap_88_89/qrel_validation", verbose=True)

Running grid eval on 10 possible combinations
Running evaluation 1 of 10, with params: {'lmbd': 0.1}
Retrieving using grid_results/jelinek_mercer_0
Running evaluation 2 of 10, with params: {'lmbd': 0.18888888888888888}
Retrieving using grid_results/jelinek_mercer_1
Running evaluation 3 of 10, with params: {'lmbd': 0.2777777777777778}
Retrieving using grid_results/jelinek_mercer_2
Running evaluation 4 of 10, with params: {'lmbd': 0.3666666666666667}
Retrieving using grid_results/jelinek_mercer_3
Running evaluation 5 of 10, with params: {'lmbd': 0.4555555555555556}
Retrieving using grid_results/jelinek_mercer_4
Running evaluation 6 of 10, with params: {'lmbd': 0.5444444444444445}
Retrieving using grid_results/jelinek_mercer_5
Running evaluation 7 of 10, with params: {'lmbd': 0.6333333333333333}
Retrieving using grid_results/jelinek_mercer_6
Running evaluation 8 of 10, with params: {'lmbd': 0.7222222222222222}
Retrieving using grid_results/jelinek_mercer_7
Running evaluation 9 of 10, with

In [31]:
def print_grid(grid):
    for params, results in grid:
        print("Params: ", params)
        print("\tResults: ", results)

In [32]:
print_grid(jelinek_mercer_grid_results)

Params:  {'lmbd': 0.1}
	Results:  {'map_cut_1000': 0.0789, 'ndcg_cut_10': 0.133, 'recall_1000': 0.2306, 'P_5': 0.12}
Params:  {'lmbd': 0.18888888888888888}
	Results:  {'map_cut_1000': 0.0794, 'ndcg_cut_10': 0.1331, 'recall_1000': 0.2344, 'P_5': 0.12}
Params:  {'lmbd': 0.2777777777777778}
	Results:  {'map_cut_1000': 0.0796, 'ndcg_cut_10': 0.1331, 'recall_1000': 0.2368, 'P_5': 0.12}
Params:  {'lmbd': 0.3666666666666667}
	Results:  {'map_cut_1000': 0.0798, 'ndcg_cut_10': 0.1331, 'recall_1000': 0.2384, 'P_5': 0.12}
Params:  {'lmbd': 0.4555555555555556}
	Results:  {'map_cut_1000': 0.0798, 'ndcg_cut_10': 0.1331, 'recall_1000': 0.2394, 'P_5': 0.12}
Params:  {'lmbd': 0.5444444444444445}
	Results:  {'map_cut_1000': 0.0799, 'ndcg_cut_10': 0.1332, 'recall_1000': 0.2395, 'P_5': 0.12}
Params:  {'lmbd': 0.6333333333333333}
	Results:  {'map_cut_1000': 0.0799, 'ndcg_cut_10': 0.1332, 'recall_1000': 0.2398, 'P_5': 0.12}
Params:  {'lmbd': 0.7222222222222222}
	Results:  {'map_cut_1000': 0.0799, 'ndcg_cut_

##### Parameter tuning for Dirichlet Prior
Dirichlet Prior (explore different values of 𝛍 [500, 1000, 1500])

In [41]:
grid_dirichlet = {
    "mu": np.arange(300, 1501, 200)
}
dirichlet_grid_results = grid_eval(grid_dirichlet, dirichlet_prior, "./ap_88_89/qrel_validation", verbose=True)

Running grid eval on 7 possible combinations
Running evaluation 1 of 7, with params: {'mu': 300}
Retrieving using grid_results/dirichlet_prior_0


KeyboardInterrupt: 

In [35]:
print_grid(dirichlet_grid_results)

Params:  {'mu': 500}
	Results:  {'map_cut_1000': 0.0838, 'ndcg_cut_10': 0.1467, 'recall_1000': 0.2513, 'P_5': 0.1333}
Params:  {'mu': 700}
	Results:  {'map_cut_1000': 0.0831, 'ndcg_cut_10': 0.1438, 'recall_1000': 0.2492, 'P_5': 0.1333}
Params:  {'mu': 900}
	Results:  {'map_cut_1000': 0.0827, 'ndcg_cut_10': 0.1389, 'recall_1000': 0.2483, 'P_5': 0.1267}
Params:  {'mu': 1100}
	Results:  {'map_cut_1000': 0.0824, 'ndcg_cut_10': 0.1369, 'recall_1000': 0.2478, 'P_5': 0.1267}
Params:  {'mu': 1300}
	Results:  {'map_cut_1000': 0.0822, 'ndcg_cut_10': 0.1371, 'recall_1000': 0.2477, 'P_5': 0.1333}
Params:  {'mu': 1500}
	Results:  {'map_cut_1000': 0.0821, 'ndcg_cut_10': 0.1342, 'recall_1000': 0.2466, 'P_5': 0.1333}


##### Parameter tuning for Absolute discounting
Absolute discounting (explore different values of 𝛅 in the range [0.1, 0.5, 0.9]). [5 points]

In [38]:
grid_abs_discounting = {
    "delta": np.linspace(0.1, 0.9, 10)
}
abs_discounting_grid = grid_eval(grid_abs_discounting, absolute_discounting, 
                                 "./ap_88_89/qrel_validation", verbose=True)

Running grid eval on 10 possible combinations
Running evaluation 1 of 10, with params: {'delta': 0.1}
Retrieving using grid_results/absolute_discounting_0
Running evaluation 2 of 10, with params: {'delta': 0.18888888888888888}
Retrieving using grid_results/absolute_discounting_1
Running evaluation 3 of 10, with params: {'delta': 0.2777777777777778}
Retrieving using grid_results/absolute_discounting_2
Running evaluation 4 of 10, with params: {'delta': 0.3666666666666667}
Retrieving using grid_results/absolute_discounting_3
Running evaluation 5 of 10, with params: {'delta': 0.4555555555555556}
Retrieving using grid_results/absolute_discounting_4
Running evaluation 6 of 10, with params: {'delta': 0.5444444444444445}
Retrieving using grid_results/absolute_discounting_5
Running evaluation 7 of 10, with params: {'delta': 0.6333333333333333}
Retrieving using grid_results/absolute_discounting_6
Running evaluation 8 of 10, with params: {'delta': 0.7222222222222222}
Retrieving using grid_results

In [ ]:
print_grid(abs_discounting_grid)

In [ ]:
def toss_coin(heads = 0.5):
    # head = True, Tails is False
    return True if random.random() < heads else False

In [ ]:
def randomiser(list_a, list_b, N = 30):
    """
    TODO
    """
    assert len(list_a) == len(list_b)
    for i in range(N):
        new_a, new_b = [], []
        
        for index in range(len(list_a)):
            if toss_coin():
                # don't swap, copy from source
                new_a.append(list_a[index])
                new_b.append(list_b[index])
            else:
                new_a.append(list_b[index])
                new_b.append(list_a[index])
        mean_new_a = np.mean(new_a)
        mean_new_b = np.mean(new_b)
        
        yield mean_new_a, mean_new_b
        
a = np.transpose(list(randomiser([1,2,3],[4,5,6])))
print(a)
z, p = stats.ttest_rel(a[0],a[1])

print(z,p)

In [ ]:
def bonferroni(groups, alpha = 0.05):
    comparisons = math.factorial(groups)/(math.factorial(2)*math.factorial(groups - 2))
    return alpha/comparisons

### Task 2: Latent Semantic Models (LSMs) [15 points] ###

In this task you will experiment with applying distributional semantics methods ([LSI](http://lsa3.colorado.edu/papers/JASIS.lsi.90.pdf) **[5 points]** and [LDA](https://www.cs.princeton.edu/~blei/papers/BleiNgJordan2003.pdf) **[5 points]**) for retrieval.

You do not need to implement LSI or LDA on your own. Instead, you can use [gensim](http://radimrehurek.com/gensim/index.html). An example on how to integrate Pyndri with Gensim for word2vec can be found [here](https://github.com/cvangysel/pyndri/blob/master/examples/word2vec.py). For the remaining latent vector space models, you will need to implement connector classes (such as `IndriSentences`) by yourself.

In order to use a latent semantic model for retrieval, you need to:
   * build a representation of the query **q**,
   * build a representation of the document **d**,
   * calculate the similarity between **q** and **d** (e.g., cosine similarity, KL-divergence).
     
The exact implementation here depends on the latent semantic model you are using. 
   
Each of these LSMs come with various hyperparameters to tune. Make a choice on the parameters, and explicitly mention the reasons that led you to these decisions. You can use the validation set to optimize hyper parameters you see fit; motivate your decisions. In addition, mention clearly how the query/document representations were constructed for each LSM and explain your choices.

In this experiment, you will first obtain an initial top-1000 ranking for each query using TF-IDF in **Task 1**, and then re-rank the documents using the LSMs. Use TREC Eval to obtain the results and report on `NDCG@10`, Mean Average Precision (`MAP@1000`), `Precision@5` and `Recall@1000`.

Perform significance testing **[5 points]** (similar as in Task 1) in the class of semantic matching methods.

In [ ]:
import gensim

In [34]:
def document_token_iterator():
    for int_doc_id in range(index.document_base(), index.maximum_document()):
        ext_doc_id, doc_token_ids = index.document(int_doc_id)
        ll = [id2token[token_id].lower() for token_id in doc_token_ids if token_id > 0]
        yield ll

In [35]:
def get_dictionary(dict_file_name = "./index.dict"):
    
    if os.path.exists(dict_file_name):
        print("Loading existing dictionary")
        return gensim.corpora.Dictionary.load(dict_file_name)
    print("Creating dictionary, and saving...")
    gensim_dict = gensim.corpora.Dictionary(document_token_iterator())
    gensim_dict.save(dict_file_name)
    
    return gensim_dict
gensim_dict = get_dictionary()

Creating dictionary, and saving...


In [36]:
gensim_dict.token2id

{'patons': 244141,
 'demeaned': 91890,
 'woundings': 71634,
 'potasinskas': 133019,
 'obiols': 247384,
 'zacharias': 87110,
 'conciliating': 204982,
 'clasmates': 176447,
 'lomaxs': 17225,
 'salerno': 43352,
 'boreanaz': 202023,
 'gerontis': 139756,
 'welcom': 153516,
 'seimograph': 31253,
 'platona': 232252,
 'awardin': 108253,
 'stray': 7610,
 'vargas': 21367,
 'stamatoiu': 164354,
 'erleen': 207865,
 'warbled': 180223,
 'bibliotheque': 47301,
 'jetprop': 97711,
 'claassen': 122820,
 'fisheries': 940,
 'humberto': 54210,
 'whisler': 210799,
 'grohses': 261725,
 'wrath': 16209,
 'mcdlts': 85797,
 'toiling': 27685,
 'observance': 23835,
 'yoshina': 208024,
 'cowsll': 180113,
 'adaption': 180794,
 'ferdovzi': 255923,
 'kanagawa': 112642,
 'kherkel': 256516,
 'pharmacists': 23516,
 'wohlin': 160288,
 'resurgency': 135919,
 '-5678': 86561,
 'boson': 195620,
 'februarys': 243,
 'hitatement': 81637,
 '-1907': 151879,
 'krasucki': 192779,
 'willye': 123235,
 'peggi': 247376,
 'wojewodski': 2

In [39]:
def get_corpus(gensim_dict, corpus_file="ir1.mm"):
    if os.path.exists(corpus_file):
        print("Loading existing corpus")
        return gensim.corpora.MmCorpus(corpus_file)
    print("Creating corpus, and saving....")
    gensim.corpora.MmCorpus.serialize(corpus_file, (gensim_dict.doc2bow(doc) for doc in document_token_iterator()))
    return gensim.corpora.MmCorpus(corpus_file)
gensim_corpus = get_corpus(gensim_dict)

Loading existing corpus


ValueError: not enough values to unpack (expected 3, got 0)

In [38]:
num_topics = 256
lsi_model = gensim.models.lsimodel.LsiModel(corpus=gensim_corpus, id2word=gensim_dict, num_topics=num_topics)

NameError: name 'gensim_corpus' is not defined

In [ ]:
lsi_model = gensim.models.lsimodel.LsiModel.load("lsi.model")

In [ ]:
lsi_model.save("lsi.model")

In [ ]:
lsi_model.show_topics()

In [ ]:
lsi_model[dictionary.doc2bow(['washington'])]

In [ ]:
from scipy.spatial.distance import cosine as cosine_distance
cosine_distance_vectorized = np.vectorize(cosine_distance)


In [ ]:
class LSMScorer:
    def __init__(self, corpus, dictionary, model):
        self.model = model
        self.corpus = corpus
        self.dict = dictionary
        
        self.document_vectors = np.zeros((num_documents, self.model.num_topics))
        self.document_index = []
        for idx, int_doc_id in enumerate(range(index.document_base(), index.maximum_document())):
            ext_doc_id, doc_token_ids = index.document(int_doc_id)
            document_tokens = [id2token[token_id].lower() for token_id in doc_token_ids if token_id > 0]
            self.document_index.append(ext_doc_id)
            self.document_vectors[idx] = self._get_vector(document_tokens)    
    
    def _get_vector(self, tokens):
        vector = np.zeros(self.model.num_topics)
        # TODO why isn't the size always = num_topics? It's super strange! 
        # TODO check back why this is happening. it's a serious problem
        for topic, value in self.model[self.dict.doc2bow(tokens)]:
            vector[topic] = value
        return vector
    
    def score(self, query_tokens):
        # TODO order = 1 same, -1 not same
        # TODO cehck others as well
        # TODO some values are 
        # TODO have to use JS divergence
        query_vector = self._get_vector(query_tokens)
        distances = np.apply_along_axis(lambda _: cosine_distance(query_vector, _), 1, self.document_vectors)
        #for ext_doc_id, document_vector in self.document_vectors.items():
        #    results.append((cosine_distance(document_vector, query_vector), ext_doc_id))
        results = list(zip(distances, self.document_index))
        results.sort(key=lambda _: -_[0])
        return results

lsi_scorer = LSMScorer(gensim_corpus, gensim_dict, lsi_model)

In [ ]:
def run_retrieval_lsm(model_name, scorer):
    """
    Runs a retrieval method for all the queries and writes the TREC-friendly results in a file.
    
    :param model_name: the name of the model (a string)
    :param scorer: the LSMScorer 
    """
    
    run_out_path = '{}.run'.format(model_name)

    #if os.path.exists(run_out_path):
    #    return

    retrieval_start_time = time.time()

    print('Retrieving using', model_name)
    
    data = {}
    start_time = time.time()
    for index, (query_id, query_tokens) in enumerate(tokenized_queries.items()):
        query_tokens = [id2token[token_id].lower() for token_id in query_tokens if token_id > 0]
        data[query_id] = lsi_scorer.score(query_tokens)
        if (index + 1) % 10 == 0:
            print("\t", index + 1 , "done.", time.time() - start_time, "elapsed" )

    with open(run_out_path, 'w') as f_out:
        write_run(
            model_name=model_name,
            data=data,
            out_f=f_out,
            max_objects_per_query=1000)


In [ ]:
run_retrieval_lsm("lsi", lsi_model)

In [ ]:
trec.evaluate("./ap_88_89/qrel_validation", "lsi.run")

### Task 3:  Word embeddings for ranking [20 points] (open-ended) ###

First create word embeddings on the corpus we provided using [word2vec](http://arxiv.org/abs/1411.2738) -- [gensim implementation](https://radimrehurek.com/gensim/models/word2vec.html). You should extract the indexed documents using pyndri and provide them to gensim for training a model (see example [here](https://github.com/nickvosk/pyndri/blob/master/examples/word2vec.py)).
   
This is an open-ended task. It is left up you to decide how you will combine word embeddings to derive query and document representations. Note that since we provide the implementation for training word2vec, you will be graded based on your creativity on combining word embeddings for building query and document representations.

Note: If you want to experiment with pre-trained word embeddings on a different corpus, you can use the word embeddings we provide alongside the assignment (./data/reduced_vectors_google.txt.tar.gz). These are the [google word2vec word embeddings](https://code.google.com/archive/p/word2vec/), reduced to only the words that appear in the document collection we use in this assignment.

In [45]:
import copy
import gensim
import logging
import pyndri
import pyndri.compat
import sys

logging.basicConfig(level=logging.INFO)

index = pyndri.Index('index/')
print(index)

print('Loading vocabulary.')
dictionary = pyndri.extract_dictionary(index)
sentences = pyndri.compat.IndriSentences(index, dictionary)

print('Initializing word2vec.')
word2vec_init = gensim.models.Word2Vec(
    size=300,  # Embedding size
    window=5,  # One-sided window size
    sg=True,  # Skip-gram.
    min_count=5,  # Minimum word frequency.
    sample=1e-3,  # Sub-sample threshold.
    hs=False,  # Hierarchical softmax.
    negative=10,  # Number of negative examples.
    iter=1,  # Number of iterations.
    workers=8,  # Number of workers.
)

print('Constructing word2vec vocabulary.')
word2vec_init.build_vocab(sentences, trim_rule=None)

models = [word2vec_init]

for epoch in range(1, 5 + 1):

    print('Epoch %d', epoch)
    model = copy.deepcopy(models[-1])
    
    '''
    Any direct calls to method train() of Word2Vec/Doc2Vec now require an explicit epochs parameter and explicit estimate of corpus size. 
    The most usual way to call train is vec_model.train(sentences, total_examples=self.corpus_count, epochs=self.iter)
    '''
    #model.train(sentences)
    model.train(sentences, total_examples=model.corpus_count, epochs=model.iter)
    models.append(model)

print('Trained models: %s', models)

with open('./ap_88_89/topics_title', 'r') as f_topics:
    queries = parse_topics([f_topics])

# the total number of documents
num_documents = index.maximum_document() - index.document_base()
# tokenize the queries
# query_id --> [token_id1, token_id2, ...] 
tokenized_queries = {
    query_id: [dictionary.translate_token(token)
               for token in index.tokenize(query_string)
               if dictionary.has_token(token)]
    for query_id, query_string in queries.items()}

# gather all of the query token_ids into a set TODO why? 
query_term_ids = set(
    query_term_id
    for query_term_ids in tokenized_queries.values()
    for query_term_id in query_term_ids)

print('Gathering statistics about', len(query_term_ids), 'terms.')

<pyndri.Index of 164597 documents>
Loading vocabulary.
Initializing word2vec.
Constructing word2vec vocabulary.


KeyboardInterrupt: 

### Task 4: Learning to rank (LTR) [15 points] (open-ended) ###

In this task you will get an introduction into learning to rank for information retrieval.

You can explore different ways for devising features for the model. Obviously, you can use the retrieval methods you implemented in Task 1, Task 2 and Task 3 as features. Think about other features you can use (e.g. query/document length). Creativity on devising new features and providing motivation for them will be taken into account when grading.

For every query, first create a document candidate set using the top-1000 documents using TF-IDF, and subsequently compute features given a query and a document. Note that the feature values of different retrieval methods are likely to be distributed differently.

You are adviced to start some pointwise learning to rank algorithm e.g. logistic regression, implemented in [scikit-learn](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html).
Train your LTR model using 10-fold cross validation on the test set. More advanced learning to rank algorithms will be appreciated when grading.

### Task 4: Write a report [15 points; instant FAIL if not provided] ###

The report should be a PDF file created using the [sigconf ACM template](https://www.acm.org/publications/proceedings-template) and will determine a significant part of your grade.

   * It should explain what you have implemented, motivate your experiments and detail what you expect to learn from them. **[10 points]**
   * Lastly, provide a convincing analysis of your results and conclude the report accordingly. **[10 points]**
      * Do all methods perform similarly on all queries? Why?
      * Is there a single retrieval model that outperforms all other retrieval models (i.e., silver bullet)?
      * ...

**Hand in the report and your self-contained implementation source files.** Only send us the files that matter, organized in a well-documented zip/tgz file with clear instructions on how to reproduce your results. That is, we want to be able to regenerate all your results with minimal effort. You can assume that the index and ground-truth information is present in the same file structure as the one we have provided.
